### PR TITLE
Add autofix to declaration-block-no-duplicate-properties

### LINKED ISSUE #6119: Add autofix to declaration-block-no-duplicate-properties
### What is the problem you're trying to solve?

I have set the Stylelint config to remove duplicate CSS properties and values as below. 

```
        "declaration-block-no-duplicate-properties": [
            true,
            {
                "ignore": ["consecutive-duplicates-with-different-values"]
            }
        ],
```
I want to remove the duplicate property values like on the fix lint command or on Ctrl+S but this is not happening at all.
Example as below.

```
.full-screen {
    align-items: center;
    align-items: center;

    //you need to setup file-loader in webpack before you can use images
    background-position: center;
    bottom: 0;
    display: flex;
    display: flex;
    flex-direction: column;

    //_ works with row or column_

    flex-direction: column;
    height: 100%;
    justify-content: center;
    justify-content: center;
    left: 0;
    position: fixed;
    right: 0;
    text-align: center;
    top: 0;
    width: 100%;
}
```

Please let me know how I can achieve this?the 

### What solution would you like to see?

I want to remove the duplicate property values like on the fix lint command and using Ctrl+S

### PR DESCRIPTION
**Pull request by fpetrakov**
> Which issue, if any, is this issue related to?

Closes #6119.

> Is there anything in the PR that needs further explanation?

No, it's self-explanatory.

### CODE REVIEW

File: lib/rules/declaration-block-no-duplicate-properties/__tests__/index.js [hunk 1]
```diff
@@ -231,6 +349,25 @@ testRule({
 	],
 });
 
+testRule({
+	ruleName,
+	config: [true, { ignore: ['consecutive-duplicates-with-different-values'] }],
+	fix: true,
+
```
Comment (Mouvedia) [inline lib/rules/declaration-block-no-duplicate-properties/__tests__/index.js:356]: Can you add an example for accept?
i.e. `display: inline-block; display: inline;` should remain `display: inline-block; display: inline;` after fix

File: lib/rules/declaration-block-no-duplicate-properties/__tests__/index.js [hunk 2]
```diff
@@ -174,6 +174,95 @@ testRule({
 	],
 });
 
+testRule({
+	ruleName,
+	config: [true],
+	fix: true,
```
Comment (ybiquitous) [inline lib/rules/declaration-block-no-duplicate-properties/__tests__/index.js:180]: [suggest] Let's merge new test cases with existing ones like this:

```diff
 testRule({
 	ruleName,
 	config: [true],
+	fix: true,
```

**Review by ybiquitous — CHANGES_REQUESTED** [review]
@fpetrakov Thanks for creating the PR! 👍🏼 

Can you please address the following suggestions?

- add `fixable: true` to the metadata like this:
  https://github.com/stylelint/stylelint/blob/b20e672dece29e50d2508be406c6f249fee9cfe1/lib/rules/alpha-value-notation/index.js#L23
- add a description to the rule readme like this:
  https://github.com/stylelint/stylelint/blob/b20e672dece29e50d2508be406c6f249fee9cfe1/lib/rules/alpha-value-notation/README.md?plain=1#L12

File: lib/rules/declaration-block-no-duplicate-properties/index.js [hunk 3]
```diff
@@ -58,9 +58,11 @@ const rule = (primary, secondaryOptions) => {
...
 		eachDeclarationBlock(root, (eachDecl) => {
-			/** @type {string[]} */
+			/** @type {import('postcss').Declaration[]} */
 			const decls = [];
 			/** @type {string[]} */
+			const declsProps = [];
+			/** @type {string[]} */
 			const values = [];
 
 			eachDecl((decl) => {
```
Comment (ybiquitous) [inline lib/rules/declaration-block-no-duplicate-properties/index.js:68]: [suggest] It seems possible to replace a new variable `decls` with the third argument of `eachDecl()` like this:

```suggestion
			eachDecl((decl, _index, decls) => {
```

Comment (ybiquitous) [reply lib/rules/declaration-block-no-duplicate-properties/index.js:68]: Ah, the new variable `decls` can be a smaller array than the third argument `decls`.
Sorry, keep the current code as-is.

File: lib/rules/declaration-block-no-duplicate-properties/index.js [hunk 4]
```diff
@@ -85,12 +87,18 @@ const rule = (primary, secondaryOptions) => {
 					return;
 				}
 
-				const indexDuplicate = decls.indexOf(prop.toLowerCase());
+				const indexDuplicate = declsProps.indexOf(prop.toLowerCase());
```
Comment (ybiquitous) [inline lib/rules/declaration-block-no-duplicate-properties/index.js:90]: [suggest] I notice we can do the following refactorings:

- Avoid repetition of `prop.toLowerCase()`
- Make `declsProps` and `values` variables unneeded by `decls.findIndex()`

For instance,

```js
const lowerProp = prop.toLowerCase();

// ...

const indexDuplicate = decls.findIndex(d => d.prop.toLowerCase() === lowerProp);

// ...

const duplicateDecl = decls[indexDuplicate];
const duplicateValue = duplicateDecl ? duplicateDecl.value : '';
```

File: lib/rules/declaration-block-no-duplicate-properties/__tests__/index.js (hunk 2 above)
Comment (fpetrakov) [reply lib/rules/declaration-block-no-duplicate-properties/__tests__/index.js:180]: > [suggest] Let's merge new test cases with existing ones like this:

You mean adding `fix: true` to the first `testRule` and replacing all rejected tests within it with fixed versions from the second `testRule`?

Comment (ybiquitous) [reply lib/rules/declaration-block-no-duplicate-properties/__tests__/index.js:180]: I mean adding `fix: true` to each `testRule`. Then, all rejected test cases should have a `fixed` property, and the diff in test cases should be fewer.

File: lib/rules/declaration-block-no-duplicate-properties/index.js (hunk 4 above)
Comment (fpetrakov) [reply lib/rules/declaration-block-no-duplicate-properties/index.js:90]: Also thought of just using decls. Great suggestion!

**Comment by fpetrakov** [comment]
Hmm, now all tests are failing, let's see...

**Comment by ybiquitous** [comment]
It seems `npx jest -u` (updating snapshots) should resolve the failures.

File: lib/rules/declaration-block-no-duplicate-properties/__tests__/index.js [hunk 5]
```diff
@@ -360,6 +398,7 @@ testRule({
 	ruleName,
 	config: [true],
 	customSyntax: 'postcss-css-in-js',
+	fix: true,
```
Comment (ybiquitous) [inline lib/rules/declaration-block-no-duplicate-properties/__tests__/index.js:401]: Ah, sorry. The cases for `postcss-css-in-js` has been skipped, so, please revert the change in this `testRule`?

File: lib/rules/declaration-block-no-duplicate-properties/__tests__/index.js [hunk 6]
```diff
@@ -377,3 +415,30 @@ testRule({
...
+				"import styled from 'styled-components';\nexport default styled.div` color: orange; `;",
+			message: messages.rejected('color'),
+		},
+		{
+			code: "import styled from 'react-emotion'\nexport default styled.div` color: pink; color: orange; `;",
+			fixed: "import styled from 'react-emotion'\nexport default styled.div` color: orange; `;",
+			message: messages.rejected('color'),
+		},
+	],
+});
```
Comment (ybiquitous) [inline lib/rules/declaration-block-no-duplicate-properties/__tests__/index.js:444]: [suggest] This code block using `testRule` is unnecessary because `skip: true` is set. Can you please remove it?

```suggestion
```

Comment (fpetrakov) [reply lib/rules/declaration-block-no-duplicate-properties/__tests__/index.js:444]: Why keeping tests that have `skip: true` btw?

Comment (ybiquitous) [reply lib/rules/declaration-block-no-duplicate-properties/__tests__/index.js]: We've been discussing deprecating the `postcss-css-in-js` package (see https://github.com/stylelint/postcss-css-in-js/issues/225).

Of course, we can remove the skipped cases (including other rules) soon, but it seems reasonable to do it with one PR.

```console
$ git grep postcss-css-in-js lib/rules | wc -l
      28
```

Also, we can do it after we determine to deprecate the package and remove it entirely from our codebase. It seems unnecessary to me to hurry.

**Review by ybiquitous — APPROVED** [review]
@fpetrakov Thank you! LGTM 👍🏼

**Comment by fpetrakov** [comment]
@ybiquitous should we merge it or call jeddy3 to review changes?

**Comment by ybiquitous** [comment]
Let's wait for a few more additional reviews. I may have missed something.

**Review by jeddy3 — APPROVED** [review]
@fpetrakov LGTM, thank you. 

@ybiquitous Thanks for the initial review and suggestions!

> should we merge it or call jeddy3 to review changes?

We usually use our judgement to determine how many reviews a pull request needs. Although, it's usually one for simple things and, as @ybiquitous suggested, two when adding new features like options or rules.

**Comment by JounQin** [comment]
If we going to merge, don't forget to add the changesets, but it's already a new iteration.

**Comment by ybiquitous** [comment]
@JounQin Thanks for the reminder. I've pushed fd25dfb. No problems?

**Comment by JounQin** [comment]
> @JounQin Thanks for the reminder. I've pushed [fd25dfb](https://github.com/stylelint/stylelint/commit/fd25dfb8555a96d903d2f56e295d994e2349c77c). No problems?

No problem.
